main/src/ca/uwaterloo/flix/language/phase/njvm/classes/TupleClass.scala

```diff
@@ -0,0 +1,242 @@
+  //Setup
+  private val ct: Reference = getTupleClassType(elms)
+  private val cg: ClassGenerator = new ClassGenerator(ct, List(getTupleInterfaceType(elms)))
+
+  //Fields
+  for ((arg, ind) <- elms.zipWithIndex) {
+      arg match {
+        case PrimBool => cg.mkPrimField[MBool]("field" + ind)
+        case PrimChar => cg.mkPrimField[MChar]("field" + ind)
+        case PrimByte => cg.mkPrimField[MByte]("field" + ind)
+        case PrimShort => cg.mkPrimField[MShort]("field" + ind)
+        case PrimInt => cg.mkPrimField[MInt]("field" + ind)
+        case PrimLong => cg.mkPrimField[MLong]("field" + ind)
+        case PrimFloat => cg.mkPrimField[MFloat]("field" + ind)
+        case PrimDouble => cg.mkPrimField[MDouble]("field" + ind)
+        case Reference(_) => cg.mkField[Ref[MObject]]("field" + ind)
+        case _ => throw InternalCompilerException(s"Unexpected type $arg")
+      }
+  }
+
+  private def getField[T1 <: MnemonicsTypes : TypeTag](ind : Int) : Field[T1] =
+    new Field[T1]("field" + ind)
+
+  private def getPrimField[T1 <: MnemonicsPrimTypes : TypeTag](ind : Int) : PrimField[T1] =
+    new PrimField[T1]("field" + ind)
+
+  val defaultConstrutor : UncheckedVoidMethod = {
+    val setFields =
+      (sig : UncheckedFunSig)  =>{
+        var ins = NO_OP[StackNil]
+        var offset: Int = 1
+        for ((arg,ind) <- elms.zipWithIndex) {
+          ins = ins |>>
+          sig.getArg[Ref[TupleClass]](0).LOAD |>>
+          (arg match {
+            case PrimBool => sig.getArg[MBool](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+                getPrimField(ind).PUT_FIELD
+            case PrimChar =>  sig.getArg[MChar](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimByte =>   sig.getArg[MByte](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimShort => sig.getArg[MShort](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimInt =>   sig.getArg[MInt](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimLong =>   sig.getArg[MLong](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimFloat =>  sig.getArg[MFloat](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case PrimDouble =>   sig.getArg[MDouble](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getPrimField(ind).PUT_FIELD
+            case Reference(_) =>   sig.getArg[Ref[MObject]](offset).LOAD[StackNil ** Ref[TupleClass]] |>>
+              getField(ind).PUT_FIELD
+            case _ => throw InternalCompilerException(s"Unexpected type $arg")
+          })
+          arg match {
+            case PrimLong | PrimDouble => offset += 2
+            case _ => offset += 1
```

Comment:
Here is a more important comment:

This whole business with computing offsets all over in the old backend was very fragile.

It would be nice if we could do something better.